### PR TITLE
Use Collections.addAll where feasible

### PR DESCRIPTION
**Pull request by OLPMO**
Simplifed some if statement and use Collection.addAll instead of iterator to improve the  efficiency

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@OLPMO Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-framework&pullRequestId=23478)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-framework&pullRequestId=23478) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@OLPMO Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-framework&pullRequestId=23478)!

File: spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java [hunk 1]
```diff
@@ -382,7 +382,7 @@ public static Method findNewInstance(Class iface) {
 			}
 		}
 		methods.remove(null);
-		return (Method[]) methods.toArray(new Method[methods.size()]);
+		return (Method[]) methods.toArray(new Method[0]);
```
Comment (mentallurg) [inline spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java]: This is bad, because it it creates **2 arrays**: first with `new Method[0]`, then `toArray()` will see that the capacity of the given array is not sufficient and will create **one more array**.

**Review by mentallurg — CHANGES_REQUESTED** [review]
These changes actually lead to extra calls for memory allocation (which means worse performance) and make the code **worse**. Besides, this code uses **more** memory than the existing code, because even an empty array needs memory for meta. I suggest to **reject** this change.

File: spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java (hunk 1 above)
Comment (sbrannen) [reply spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java]: I agree with @mentallurg here.

Please revert this change.

File: spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java [hunk 2]
```diff
@@ -412,7 +412,7 @@ public static Method findNewInstance(Class iface) {
 					properties.add(pd);
 				}
 			}
-			return (PropertyDescriptor[]) properties.toArray(new PropertyDescriptor[properties.size()]);
+			return (PropertyDescriptor[]) properties.toArray(new PropertyDescriptor[0]);
```
Comment (sbrannen) [inline spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java]: Please revert this change.

File: spring-core/src/main/java/org/springframework/core/annotation/AnnotationsScanner.java [hunk 3]
```diff
@@ -234,9 +234,7 @@ private AnnotationsScanner() {
 		if (includeEnclosing && enclosingClass != null) {
 			R enclosingResult = processClassHierarchy(context, aggregateIndex,
 					enclosingClass, processor, classFilter, includeInterfaces, true);
-			if (enclosingResult != null) {
-				return enclosingResult;
-			}
+			return enclosingResult;
```
Comment (sbrannen) [inline spring-core/src/main/java/org/springframework/core/annotation/AnnotationsScanner.java]: Although this may be technically equivalent, it breaks with the structure of the rest of the method.

As such, I do not think the change is warranted.

So, please revert this change.

File: spring-core/src/main/java/org/springframework/core/annotation/AnnotationsScanner.java [hunk 4]
```diff
@@ -323,9 +321,7 @@ private AnnotationsScanner() {
 		if (superclass != Object.class && superclass != null) {
 			R superclassResult = processMethodHierarchy(context, aggregateIndex,
 					superclass, processor, classFilter, rootMethod, includeInterfaces);
-			if (superclassResult != null) {
-				return superclassResult;
-			}
+			return superclassResult;
```
Comment (sbrannen) [inline spring-core/src/main/java/org/springframework/core/annotation/AnnotationsScanner.java]: Although this may be technically equivalent, it breaks with the structure of the rest of the method.

As such, I do not think the change is warranted.

So, please revert this change.

**Review by sbrannen — CHANGES_REQUESTED** [review]

File: spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java (hunk 1 above)
Comment (andrei-ivanov) [reply spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java]: What about https://shipilev.net/blog/2016/arrays-wisdom-ancients/ ?

Comment (OLPMO) [reply spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java]: @andrei-ivanov The information you provided is much more detailed than mine. I have learned a lot from it, thank you.

Comment (sbrannen) [reply spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java]: Indeed. I had momentarily forgotten about those findings from Shipilёv.

Thanks for the links.

We'll keep the proposed change! 👍

File: spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java (hunk 2 above)
Comment (sbrannen) [reply spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java]: See previous comments.

**Review by sbrannen — APPROVED** [review]

File: spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java (hunk 1 above)
Comment (sbrannen) [inline spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java]: Please revert this change.

Rationale:

_Please refrain from modifying classes under `org.springframework.asm`, `org.springframework.cglib`, and `org.springframework.objenesis`. Those include repackaged forks of the third-party libraries ASM, CGLIB, and Objenesis. Any refactoring to those classes should take place upstream in the originating repository. The Spring Framework will then pick up the changes when syncing with official updates of the forked third-party libraries._

**Review by sbrannen — CHANGES_REQUESTED** [review]
The changes look good now.

I have only asked for the change to the repackaged CGLIB code be reverted due to our policy for repackaged third-party libraries.

File: spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java (hunk 1 above)
Comment (OLPMO) [reply spring-core/src/main/java/org/springframework/cglib/core/ReflectUtils.java]: OK, thank you for your analysis. @sbrannen

**Review by sbrannen — APPROVED** [review]

**Comment by sbrannen** [comment]
This has been merged into `master`.

Thanks

**Comment by OLPMO** [comment]
It's my pleasure.
